rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -0,0 +1,261 @@
+                to_counterparty=CPT_RAINBOW_SWAPS,
+                to_notes=f'Receive {amount} {in_asset.resolve_to_asset_with_symbol().symbol} as the result of a Rainbow Swap',  # noqa: E501
+                address=self.router_address,
+            ))
+
+        return DecodingOutput(event=fee_event)
+
+    def _decode_swap_native_to_token(self, context: DecoderContext) -> DecodingOutput:
+        """
+        Decodes a swap transaction from native token to another token.
+        This function the fee charged using the wrapperd nativa token deposit log
+        and the token being bought in a swap transaction where the native token is swapped for another token.
+        """  # noqa: E501
+
+        buy_asset = None
+        fee_raw = None
+        buy_amount = None
+        for log in context.all_logs:
+            # Find the fee charged using the deposit log on wrapped native token contract
+            if log.topics[0] == WETH_DEPOSIT_TOPIC:
+                fee_raw = context.transaction.value - int.from_bytes(log.data)
+            # Find the buy token using the transfer log
+            if log.topics[0] == ERC20_OR_ERC721_TRANSFER and bytes_to_address(log.topics[2]) == context.transaction.from_address:  # noqa: E501
+                buy_asset = self.base.get_or_create_evm_token(log.address)
+                buy_amount = int.from_bytes(log.data)
+
+            if buy_asset and buy_amount and fee_raw:
+                break
+
+        if not (buy_asset and fee_raw and buy_amount):
+            return DEFAULT_DECODING_OUTPUT
+
+        return self._decode_swap(
+            context=context,
+            fee_raw=fee_raw,
+            in_asset=buy_asset,
+            in_amount=buy_amount,
+        )
+
+    def _decode_swap_token_to_native(self, context: DecoderContext) -> DecodingOutput:
+        tx_log = context.tx_log
+        sender = context.transaction.from_address
+        native_token = self.evm_inquirer.native_token
+        if bytes_to_address(tx_log.topics[2]) == self.router_address:
+            in_event = None
+            fee_raw = None
+            for event in context.decoded_events:
+                # FCind the receive event
+                if (event.event_type == HistoryEventType.RECEIVE and event.event_subtype == HistoryEventSubType.NONE and event.location_label == sender):  # noqa: E501
+                    event.event_type = HistoryEventType.TRADE
+                    event.event_subtype = HistoryEventSubType.RECEIVE
+                    event.counterparty = CPT_RAINBOW_SWAPS
+                    event.notes = f'Receive {event.amount} {event.asset.resolve_to_asset_with_symbol().symbol} as the result of a Rainbow Swap'  # noqa: E501
+                    event.address = self.router_address
+                    in_event = event
+
+                    # Find the log withdraw the weth token
+                    for log in context.all_logs:
+                        if log.topics[0] == WETH_WITHDRAW_TOPIC:
```

Comment:
as a rule if you are looping and trying to find something when you find it break the loop so you don't waste time. Also in python loops have an `else` statement where you can log if you didn't find what you were loooking for